fix(user-profile): enable saving empty bio to clear user profile field

Curious-Goblin opened the pull request:
### Proposed changes
This pull request enables the ability to save an empty bio in the user profile to clear out the bio field, which was previously uneditable if left empty.

- This fix addresses an issue where users could not save changes if they attempted to clear their bio.
- Implements validation adjustments to allow empty strings in the bio field.

### Issue(s)
- Fixes #33783

### Steps to test or reproduce
1. Navigate to the user profile section in the Rocket.Chat application.
2. Attempt to clear the bio field and save changes.
3. Confirm that the profile successfully updates with an empty bio field.

### Checklist
- [x] I have read the [Contributing Guide](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat).
- [x] I have signed the CLA.
- [x] Lint and unit tests pass locally.
- [x] Added tests to verify this fix.
- [x] Updated documentation as needed.

Gustrb commented on .changeset/slimy-lamps-learn.md:
```diff
@@ -0,0 +1,5 @@
+---
+'@rocket.chat/meteor': major
```
this is not a major change

Curious-Goblin replied:
Thank you for your feedback. I initially marked it as a major change because it involved modifying the behavior around setting an empty bio field. However I understand that this is more of  a small adjustment, so I have updated the changeset to reflect a patch version instead in the commit  #[bf964f3](https://github.com/RocketChat/Rocket.Chat/pull/33927/commits/bf964f372f1764ea526d0db30d3cde27a4c1ea06)

Gustrb commented on apps/meteor/server/methods/saveUserProfile.ts:
```diff
@@ -79,7 +79,7 @@ async function saveUserProfile(
 	}
 
-	if (user && settings.bio) {
+	if (user && settings.bio !== undefined) {
```
Could you add tests to ensure this behavior?

Curious-Goblin replied:
Thanks for the suggestion. I have added the test case to check the setting of an empty bio field as requested in the commit #bf964f3 .Please let me know if there’s anything else you'd like me to adjust.

debdutdeb replied:
I was thinking about this rather?
```suggestion
	if (user) {
```

the very next line validates string or not already.

Curious-Goblin replied:
Thank you for the suggestion. I think keeping the settings.bio !== undefined check in place allows users to have more control by ensuring that the bio field is only updated when explicitly provided. This approach prevents any unintentional updates, such as accidentally clearing the bio if the field is omitted from the request. Let me know if you’d like me to adjust this further.

Gustrb replied:
I don't think this matters since we validate it anyway, I agree with @debdutdeb on this one

debdutdeb commented on apps/meteor/app/lib/server/functions/getAvatarSuggestionForUser.ts:
```diff
@@ -53,7 +53,7 @@ const avatarProviders = {
 
 	twitter(user: IUser) {
-		if (user.services?.twitter?.profile_image_url_https && settings.get('Accounts_OAuth_Twitter')) {
+		if (user.services?.twitter?._image_url_https && settings.get('Accounts_OAuth_Twitter')) {
```
not related. please remove this change

Curious-Goblin replied:
Thank you for catching that. I have removed the unrelated changes, and the commit now only includes the intended updates for this issue.

debdutdeb reviewed (changes requested):
there should be api test for this.

Gustrb commented:
Hey, I agree with @debdutdeb we should have an API test for this as well

Gustrb commented on apps/meteor/tests/unit/server/users/saveUserIdentity.spec.ts:
```diff
@@ -131,4 +131,15 @@ describe('Users - saveUserIdentity', () => {
 		expect(result).to.be.true;
 	});
+
+	it('should update bio to an empty string', async () => {
```
I think you should ensure setBio was called, right?

Curious-Goblin replied:
Yes, I missed that, I've updated the test to include a check that ensures setBio is called when an empty bio field is set. The stub for setBio has been added to the test, and I've verified that it's invoked correctly. Let me know if anything else needs to be adjusted!"

Curious-Goblin commented:
> Hey, I agree with @debdutdeb we should have an API test for this as well



> Hey, I agree with @debdutdeb we should have an API test for this as well

Hi @Gustrb ,

I’ve added API tests for setting an empty bio field in the user profile. These tests include:

1. Validating that the saveUserProfile method allows clearing the bio by setting it to an empty string (bio: ' ').
2.  Ensuring compliance with the bio length constraints while verifying the behavior for an empty bio input.

To the best of my knowledge, these cover the requested scenarios. Could you please review and let me know if they align with the tests you asked about? If not, I’d appreciate it if you could clarify which specific API tests are referred to here so I can address them.

Thank you!

Gustrb commented:
Hey @Curious-Goblin no need to keep merging develop into your branch, you're good.
Let's run the CI to see, but looks good to me

Curious-Goblin commented on apps/meteor/tests/unit/server/users/saveUserIdentity.spec.ts:
```diff
@@ -131,4 +132,17 @@ describe('Users - saveUserIdentity', () => {
+		expect(stubs.setRealName.called).to.be.false;
+		expect(stubs.setUsername.called).to.be.false;
+		expect(stubs.updateUsernameOfEditByUserId.called).to.be.false;
+		expect(stubs.setBio.calledWith('valid_id', '')).to.be.true;
```
Hi @Gustrb ,

I tried running the tests using yarn testunit, but I wasn’t able to reproduce the failure for the above test. Could you please guide me on the steps or the specific command you used to encounter this failure? If there’s a different setup or additional dependencies required, let me know so I can address the issue effectively.

Thanks in advance for your help.

Gustrb replied:
Hey, it is failing on the CI, to run this specific test locally run
```bash
$ yarn .testunit:server -g saveUserIdentity
```

Curious-Goblin replied:
Ok, thanks

Curious-Goblin replied:
Hi @Gustrb,

I wanted to clarify a small confusion I had earlier. Initially, I was writing tests for the saveUserProfile method in a file that was intended for testing the saveUserIdentity method. However, after reviewing the test structure, I realized that there were no existing tests for the saveUserProfile method. As a result, I created a new test file specifically for it.

Although I added the new test file path to the .mocharc.js configuration so that it would be included in the test execution, I'm still facing an issue where the test is not being picked up during execution. I’m unable to run the test file, and I was hoping you could guide me on how to resolve this issue.

Could you please let me know the correct approach to ensure that this test file is included and executed?

Thanks in advance.

Gustrb replied:
Hey, I am pretty happy with the API test, from my side there is no need to add unit tests
Also, there is no need to add that path into .mocharc

Curious-Goblin replied:
Thanks for the clarification, gustrb! Just to confirm, is it necessary to have a unit test specifically for the saveUserProfile method, or is the API test sufficient to cover this functionality?

If no unit test is needed, should I go ahead and create a new commit to remove the changes related to the test file and .mocharc.js adjustments?

Let me know how you’d prefer me to proceed!

Gustrb replied:
the API test is sufficient :)
I would remove those changes

Gustrb replied:
it will be probably tested by someone else so we can get the QA tested tag
I'll approve once you change that, but there might be more reviews from other people

Curious-Goblin replied:
> the API test is sufficient :) I would remove those changes

Ok, but I kindly request that you remove these changes( the saveUserProfile file and the path in mocharc.js file) so that it does not cause any error.

Gustrb replied:
I can't do that right now, I thought you were working on that(?)
By I would I mean, "if I were you" :)

Curious-Goblin replied:
Now that I understand the expectations better, I have removed the unintended parts of the code, including the unnecessary test file and its associated configuration. The remaining changes are ready for further testing.

Curious-Goblin commented:
Hi @Gustrb,

I hope you're doing well! It's been a while since my last commit. I wanted to ask if you've had a chance to discuss it with the QA team. If there are any errors or improvements needed, please let me know—I’d be happy to address them promptly. 

Thanks for your time and guidance!

Curious-Goblin commented:
Hi @Gustrb ,
I wanted to follow up on this PR since it's been over a week without a response. Please let me know if any further changes or clarifications are needed from my side to move this forward. I’m happy to address any feedback

casalsgh commented:
Asking Gustavo to check. He had some other pressing items and had to delay this for a bit

Curious-Goblin commented:
hello @Gustrb @casalsgh, its been another week, can someone please look into this pr and see if it is good to be merged or please help to get this pr a QA assured tag so that it can be merged or if is has some issues then let me know, I would be more than happy to contribute.

reetp commented:
> hello @Gustrb @casalsgh, its been another week, can someone please look into this pr and see if it is good to be merged or please help to get this pr a QA assured tag so that it can be merged or if is has some issues then let me know, I would be more than happy to contribute.

Please note.

Nothing around here happens fast. It can take MONTHS.

There have also been the Christmas holidays which for many is up to two weeks and has only just finished.

It is with the team. You just have to be extremely patient.

Gustrb commented:
Hey, we don't merge PR's with failing tests, please fix that, then I will look into QA-ing

Curious-Goblin commented:
regarding the failing tests thought that these tests which are failing I did not think that they were not the results on my changes but I will now look into my pr and try to change things accordingly so that the tests not fail

Gustrb commented on apps/meteor/server/methods/saveUserProfile.ts:
```diff
@@ -79,7 +79,7 @@ async function saveUserProfile(
 
-	if (user && settings.bio) {
+	if (user && settings.bio !== undefined) {
 		if (typeof settings.bio !== 'string') {
```
you can probably pull this validation from this if

```ts
if (user && settings.bio !== 'string') {
    throw new Meteor.Error('error-invalid-field', 'bio', { method: 'saveUserProfile', });
}
```

Gustrb commented on apps/meteor/server/methods/saveUserProfile.ts:
```diff
@@ -79,7 +79,7 @@ async function saveUserProfile(
 	}
 
-	if (user && settings.bio) {
+	if (user && settings.bio !== undefined) {
```
I guess that if you change this to != it will work. let javascript coerce null to undefined
(empty string does not coerce to undefined)

Curious-Goblin replied:
Ok, I will apply these changes

Curious-Goblin replied:
can you please give the approval to run the workflows ??

Curious-Goblin replied:
please run the tests once again to see whether the tests are failing of not

Curious-Goblin commented:
> The UI tests are not, but there is one API test that is: `should reject invalid bio values (e.g., null):`

I think that now this error should be resolved now, can you check on it please

Gustrb commented:
the api tests are still failing

Curious-Goblin commented:
> The UI tests are not, but there is one API test that is: `should reject invalid bio values (e.g., null):`

how can I see this message in the workflows I am not able find it out, can you help me ?

Gustrb commented:
to run the api tests (the ones that are failing you can do the following:
- Open two terminals, I'll call them A and B
- On terminal A: navigate to the RC folder and run: `TEST_MODE=true yarn dsv`
- On terminal B: navigate to the RC folder and run `yarn testapi`

Note: only run the tests when the workspace is up and running

Curious-Goblin commented:
> to run the api tests (the ones that are failing you can do the following:
> 
>     * Open two terminals, I'll call them A and B
> 
>     * On terminal A: navigate to the RC folder and run: `TEST_MODE=true yarn dsv`
> 
>     * On terminal B: navigate to the RC folder and run `yarn testapi`
> 
> 
> Note: only run the tests when the workspace is up and running

Thank you very much for this guidance, I have now checked that the test which you told about, is passing successfully. Actually the problem was in the test, it was sending request to wrong api endpoint, upon fixing that the test passed without changing anything the saveUserProfile file.

Please give the approval to the workflows and see if anything more is required to be worked on.

Curious-Goblin commented:
can you please give the approval to run the workflows ??

reetp commented:
> can you please give the approval to run the workflows ??

> > This workflow requires approval from a maintainer.   <<<<<<<<<<<<<<

Please read the link above:
[Learn more about approving workflows.](https://docs.github.com/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks)

> Approving workflow runs from public forks
> When an outside contributor submits a pull request to a public repository, a maintainer with write access may need to approve some workflow runs.

The answer will be no because only rocket devs have permissions to do this. Please stop asking.  Thanks.

Curious-Goblin commented:
> > can you please give the approval to run the workflows ??
> 
> > > This workflow requires approval from a maintainer.   <<<<<<<<<<<<<<
> 
> Please read the link above: [Learn more about approving workflows.](https://docs.github.com/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks)
> 
> > Approving workflow runs from public forks
> > When an outside contributor submits a pull request to a public repository, a maintainer with write access may need to approve some workflow runs.
> 
> The answer will be no because only rocket devs have permissions to do this. Please stop asking. Thanks.

Gustrb has given the approvals for workflows for a several times and therefore I was asking to him so that we can move forward with the PR

reetp commented:
> Gustrb has given the approvals for workflows for a several times and therefore I was asking to him so that we can move forward with the PR

Ok. In which case you may need to be patient - many of the main devs do not work on the weekend.

Curious-Goblin commented:
ok

Curious-Goblin commented:
> > to run the api tests (the ones that are failing you can do the following:
> > ```
> > * Open two terminals, I'll call them A and B
> > 
> > * On terminal A: navigate to the RC folder and run: `TEST_MODE=true yarn dsv`
> > 
> > * On terminal B: navigate to the RC folder and run `yarn testapi`
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Note: only run the tests when the workspace is up and running
> 
> Thank you very much for this guidance, I have now checked that the test which you told about, is passing successfully. Actually the problem was in the test, it was sending request to wrong api endpoint, upon fixing that the test passed without changing anything the saveUserProfile file.
> 
> Please give the approval to the workflows and see if anything more is required to be worked on.

Hello gustrb, I request you to give the approval for running the workflows as on my system the test passed

Curious-Goblin commented:
I have deleted the changes in the file.

But I made those changes in order to run the test which was written by me, I ran different commands written in the script but got that dependencies are missing like the one in the ss
![Screenshot from 2025-01-13 18-36-41](https://github.com/user-attachments/assets/d550b25c-0a8c-4da5-a482-5438f9a79b21)

To overcome I installed required some dependencies locally and globally and then I reached this command
`TS_NODE_COMPILER_OPTIONS='{"module": "commonjs"}' yarn mocha tests/end-to-end/api/users.ts --config .mocharc.api.js
` which successfully ran my test but now as I reverted the changes to original this shows this error
![Screenshot from 2025-01-13 18-39-53](https://github.com/user-attachments/assets/85997289-dc4a-4ac3-8a78-3fa2173ac18c)

Maybe I am wrong, I don't know the proper way to run my test.

I was doing as you directed but got some errors

Curious-Goblin commented:
Gustrb, can you please see the latest changes and give the approval to the workflows so that we can see what more is required.

Curious-Goblin commented:
can someone please give the approval to the workflows and review this PR ??

Curious-Goblin commented:
gustrb plese take a look in this

Curious-Goblin commented:
gustrb can you please give a look to this

dougfabris commented:
Hi there, thanks for the contribution! 🚀 💯 

Closing this PR because it's stale now and it was already fixed here #37430

---

**Questions? Help needed? Feature Requests?** 

- Join our [Open Server](https://open.rocket.chat/) in the #support channel and feel free to raise a question
- Join our [Community Forum](https://forums.rocket.chat/) and search/create a post there
